src/main/java/io/cdap/plugin/http/source/common/pagination/page/BasePage.java

```diff
@@ -42,14 +56,78 @@ public BasePage(BaseHttpSourceConfig config, HttpErrorHandler httpErrorHandler,
 
   @Nullable
   public abstract String getPrimitiveByPath(String path);
+  public abstract boolean hasNext();
+  protected abstract StructuredRecord next() throws TransformationException;
```

Comment:
TransformationException here is being used like a return value. It's better if the actual implementation creates the InvalidEntry and calls the consumer rather than using the exception to control the flow of the logic. Different implementations may also want to include different fields in the error.

I think it's simpler to make consume() abstract and have the children implement it. It would also make the logic in things like RecordPerLinePage simpler, as reading through an InputStream is more straightforward than implementing an iterator over a stream.

Reply:
Unfortunately we need Page to be an iterator.

The reason is that in BasePaginationIterator we need to check page.hasNext(). And if the page does not have elements the pagination iterator needs to stop loading pages.

Context: A lot of paginated APIs return endless amount of empty pages once all elements run out.